tendermint/src/rpc/event_listener.rs

```diff
@@ -0,0 +1,292 @@
+struct TxValue {
+    #[serde(rename = "TxResult")]
+    tx_result: TxResult,
+}
+/// Tx Result
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct TxResult {
+    height: String,
+    index: i64,
+    tx: String,
+    result: TxResultResult,
+}
+/// TX Results Results
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct TxResultResult {
+    log: String,
+    gas_wanted: String,
+    gas_used: String,
+    events: Vec<TmEvent>,
+}
+
+/// Tendermint ABCI Events
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct TmEvent {
+    #[serde(rename = "type")]
+    event_type: String,
+    attributes: Vec<Attribute>,
+}
+/// Event Attributes
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct Attribute {
+    key: String,
+    value: String,
+}
+
+///Block Results JSON PRC response
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct JSONRpcBlockResult {
+    jsonrpc: String,
+    id: String,
+    result: BlockResult,
+}
+/// Block Results
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BlockResult {
+    query: String,
+    data: BlockResultData,
+    events: HashMap<String, Vec<String>>,
+}
+/// Block Results data
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BlockResultData {
+    #[serde(rename = "type")]
+    data_type: String,
+    value: BlockValue,
+}
+///Block Value
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BlockValue {
+    block: Block,
+    result_begin_block: ResultBeginBlock,
+    result_end_block: ResultEndBlock,
+}
+/// Block
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Block {
+    header: Header,
+    data: BlockData,
+    evidence: Evidence,
+    last_commit: LastCommit,
+}
+///Block Txs
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BlockData {
+    txs: Option<serde_json::Value>,
+}
+///Tendermint evidence
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Evidence {
+    evidence: Option<serde_json::Value>,
+}
+/// Header
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Header {
+    version: Version,
+    chain_id: String,
+    height: String,
+    time: String,
+    last_block_id: BlockId,
+    last_commit_hash: String,
+    data_hash: String,
+    validators_hash: String,
+    next_validators_hash: String,
+    consensus_hash: String,
+    app_hash: String,
+    last_results_hash: String,
+    evidence_hash: String,
+    proposer_address: String,
+}
+///Block ID
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BlockId {
+    hash: String,
+    parts: Parts,
+}
+/// Block Parts
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Parts {
+    total: String,
+    hash: String,
+}
+///Block version
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Version {
+    block: String,
+    app: String,
+}
+///Perevious Commit
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct LastCommit {
+    height: String,
+    round: String,
+    block_id: BlockId,
```

Comment:
These duplicate existing types. I think for the sake of getting this merged in quickly, we can leave it as is, but this code should certainly reuse the existing types as much as possible.

Reply:
Oh DRYing these should be fast

Reply:
I just pushed DRYing the ones that seemed obvious. It looks like some of the response structure here is different enough from the BlockResponse that it requires a new struct.

Reply:
Looks good to me 👍

Reply:
I further removed Block, BlockData, Evidence, Parts in 3269d58